internal/middleware/perfinsights/perfinsights.go

```diff
@@ -0,0 +1,189 @@
+package perfinsights
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/ccoveille/go-safecast"
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/rs/zerolog/log"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+
+	"github.com/authzed/ctxkey"
+	"github.com/authzed/grpcutil"
+
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+)
+
+type APIShapeLabel string
+
+const (
+	ResourceTypeLabel     APIShapeLabel = "resource_type"
+	ResourceRelationLabel APIShapeLabel = "resource_relation"
+	SubjectTypeLabel      APIShapeLabel = "subject_type"
+	SubjectRelationLabel  APIShapeLabel = "subject_relation"
+	NameLabel             APIShapeLabel = "name"
+	FilterLabel           APIShapeLabel = "filter"
+)
+
+var allLabels = []string{
+	string(ResourceTypeLabel),
+	string(ResourceRelationLabel),
+	string(SubjectTypeLabel),
+	string(SubjectRelationLabel),
+	string(NameLabel),
+	string(FilterLabel),
+}
+
+// APIShapeLabels is a map of APIShapeLabel to any value.
+type APIShapeLabels map[APIShapeLabel]any
+
+// NoLabels returns an empty APIShapeLabels map.
+func NoLabels() APIShapeLabels {
+	return APIShapeLabels{}
+}
+
+// APIShapeLatency is the metric that SpiceDB uses to keep track of the latency
+// of API calls, by shape.
+var APIShapeLatency = promauto.NewHistogramVec(prometheus.HistogramOpts{
+	Namespace:                   "spicedb",
+	Subsystem:                   "perf_insights",
+	Name:                        "api_shape_latency_seconds",
+	Help:                        "The latency of API calls, by shape",
+	Buckets:                     nil,
+	NativeHistogramBucketFactor: 1.1,
```

Comment:
Excited to see the first use of native histogram buckets! Please describe the motivation in this PR. I suspect it has to do with the cardinalities.

Note to readers: it's still [marked as "highly experimental"](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms), and it's only supported from Prometheus 2.40

Also note [a flag has to be enabled in prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms)

See more at https://prometheus.io/docs/specs/native_histograms/

Reply:
That doc is out of date -- they've been stable since 3.0

Reply:
I think this feature has been stable in Go since 3.0, but because it's not ubiquitous in all of the other languages, they've held off marking it such.

Reply:
I had to add the flag on the latest docker image to use it yesterday